### PR TITLE
finally updating the rate limits docs to be more helpful

### PR DESCRIPTION
**Pull request by tylernix**
"I'm a developer. I'd want all of this public so I don't have to talk to our support." - @aren55555

https://tryfinch.slack.com/archives/C03C8TLLFFU/p1683747754746009

### CODE REVIEW

File: docs/Development-Guides/Rate-Limits.md [hunk 1]
```diff
@@ -29,3 +22,166 @@ Product | Max requests initiated per minute
 `employment` | 4
 `payment` | 2
 `pay-statement` | 2
+
+## Application Rate Limits
+
+Similarly, multiple access tokens can be created from a single Finch application as more employers are connected. A Finch application has its own rate limits separate from the access token rate limits. You can think of an application like a development environment or application stage. Finch will assign three types of applications: `sandbox`, `development`, and `production`. Each will have their own `client_id` which represents the application.
```
Comment (aren55555) [inline docs/Development-Guides/Rate-Limits.md:28]: I'd omit the part about the kinds/types of applications, seems irrelevant to the rate limiting. I'd hope a developer knows their Client ID. I think that's a fairly common pattern.

File: docs/Development-Guides/Rate-Limits.md [hunk 2]
```diff
@@ -29,3 +22,166 @@ Product | Max requests initiated per minute
...
+
+Type | Max requests | Duration | Penalty
+-------|-------------|-------|-------------
+`API` | 1000 | 5 minutes | 60 minutes
+
+## Other types of rate limits
+
+### Finch Connect Rate limits
+
+Finch enforces rate limits on Employers going thru [Finch Connect](/docs/Integrating-with-Finch/Integrate-Finch-Connect/Overview.md). Each "step" in Finch Connect (the provider selection, logging in, multi-factor authentication prompts, etc) makes a single request to the Finch `/auth/authorize` endpoint. It is highly unlikely for an Employer to encounter these rate limit under normal use. However, an example of this rate limit being encountered is if an employer forgot their password and tried 30 different passwords in the time period of 1 minute. In this case, they would receive a rate limit penalty of 5 minutes.
```
Comment (aren55555) [inline docs/Development-Guides/Rate-Limits.md:53]: Our Finch Connect React app (in Netlify) is the only client of `/auth/authorize` -- an external developer would never call this directly. It would make sense to omit this detail here (just adds confusion).

File: docs/Development-Guides/Rate-Limits.md [hunk 3]
```diff
@@ -29,3 +22,166 @@ Product | Max requests initiated per minute
...
+
+Type | Max requests | Duration | Penalty
+-------|-------------|-------|-------------
+`Auth` | 30 | 1 minute | 5 minutes
+
+### Upstream Provider Rate Limits
+
+Because Finch integrates directly with 180+ providers, we take extreme care to avoid provider-specific rate limits. However, it inevitably happens. In this case, we will return a specific error code [upstream_rate_limit_exceeded](/docs/Development-Guides/Errors/Error-Types.md#error-types-1) explaining that Finch encountered a provider-specific rate limit. In these cases, wait a few moments before retrying. It is difficult to recommend a wait duration since every provider is different.
```
Comment (aren55555) [inline docs/Development-Guides/Rate-Limits.md:61]: No action item here -- but wanted to note this is an added benefit of getting entirely off of supporting "live" requests. These 429s are encountered by our job which has a greater "patience" to be able to slow up and retry. Even if our job can't break free of the rate-limiting, it would just end up as a failed job instead of being an error handled by the developer's code.

File: docs/Development-Guides/Rate-Limits.md [hunk 4]
```diff
@@ -29,3 +22,166 @@ Product | Max requests initiated per minute
...
+
+## Scenarios
+
+### Scenario 1: Hitting access token level rate limits
+
+In this scenario, let's focus on a single access token (Token A) and assume you are making API requests to any of the the `company`, `directory`, `individual`, `employment`, `payment`, and `pay-statement` endpoints. When a request  is sent to an endpoint, a single gallon of water is added to the endpoint’s bucket each time. The first requests starts that bucket’s 60-second Time-To-Live (TTL) timer.
+
+1. Token A makes `4` requests to the `directory` endpoint within a minute pouring 4 gallons of water into the “directory” bucket.
```
Comment (aren55555) [inline docs/Development-Guides/Rate-Limits.md:69]: Instead of "`4`" I'd write out 4 and reserve the code fence for actual code (same for the other numbers below.

Comment (aren55555) [inline docs/Development-Guides/Rate-Limits.md:69]: Instead of the shorthand "directory" I'd give the long hand GET /employer/directory (or whatever the path is)... also you could link to those docs ;)

File: docs/Development-Guides/Rate-Limits.md [hunk 5]
```diff
@@ -29,3 +22,166 @@ Product | Max requests initiated per minute
...
+        `pay-statement` | 0/2
+    - Application-level rate limits
+        Bucket | Capacity
+        -------|-------------
+        `company` | 20/20 (FULL)
+        `directory` | 16/20
+        `individual` | 0/20
+        `employment` | 0/20
+        `payment` | 12/12 (FULL)
+        `pay-statement` | 0/12
```
Comment (aren55555) [inline docs/Development-Guides/Rate-Limits.md:187]: This is very long-winded and more of a step-by-step example rather than a doc. Should this instead go on a new page?

Comment (aren55555) [reply docs/Development-Guides/Rate-Limits.md:187]: Also @joshuakrohn an image may really really help us convey this information here.

**Review by aren55555 — APPROVED** [review]

File: docs/Development-Guides/Rate-Limits.md (hunk 1 above)
Comment (tylernix) [reply docs/Development-Guides/Rate-Limits.md:42]: Unfortunately, some developers need this explained, so I'll keep it unless anyone else opposes it.

File: docs/Development-Guides/Rate-Limits.md (hunk 2 above)
Comment (tylernix) [reply docs/Development-Guides/Rate-Limits.md]: Removed

File: docs/Development-Guides/Rate-Limits.md (hunk 3 above)
Comment (tylernix) [reply docs/Development-Guides/Rate-Limits.md]: Removed

File: docs/Development-Guides/Rate-Limits.md (hunk 4 above)
Comment (tylernix) [reply docs/Development-Guides/Rate-Limits.md]: Linked to our API reference + updated links to /employer/directory

File: docs/Development-Guides/Rate-Limits.md (hunk 5 above)
Comment (tylernix) [reply docs/Development-Guides/Rate-Limits.md:215]: I agree that an image would be better, but working with what we have at the moment.

Comment (tylernix) [reply docs/Development-Guides/Rate-Limits.md:215]: @logangingerich where do you think this could go instead?

File: docs/Development-Guides/Rate-Limits.md (hunk 1 above)
Comment (aren55555) [reply docs/Development-Guides/Rate-Limits.md:42]: Maybe it should go in its own document/page explaining the Client ID differences?

File: docs/Development-Guides/Rate-Limits.md (hunk 5 above)
Comment (logangingerich) [reply docs/Development-Guides/Rate-Limits.md:215]: I had been thinking about it being a best practice doc. Alternatively, if we don't want to keep piling guides up, we could but it out as a blog post so it's at least easily shareable whenever folks need something more in-depth

Comment (tylernix) [reply docs/Development-Guides/Rate-Limits.md]: Refactored. I removed the sections and added a "Handling Rate Limit Errors" Best Practice guide.

File: docs/Development-Guides/Rate-Limits.md [hunk 6]
```diff
@@ -1,25 +1,18 @@
 # Rate Limits
 
-Finch will return a rate limit error with the HTTP status code `429` when the request rate limit for an application or `token` has been exceeded for a given `product`.
+Finch will return a rate limit error with the HTTP status code 429 when the request rate limit for an application, access token, or an individual IP address has been exceeded.
 
-*** 
+Finch's rate limits work on a per-product basis for both applications and access tokens. Rate limits are summed on a rolling 60-second basis for each unique `product`. This is commonly referred to as a Sliding or Rolling Window rate limit.
 
-## Application
+You can think of a `product` rate limit like a "bucket". Therefore, when a request is made to a `product` (which correspond directly to an API endpoint), a single gallon of water is added to that endpoint’s bucket, thus starting that bucket's 60-second Time-To-Live (TTL) timer.
```
Comment (alexromano) [inline docs/Development-Guides/Rate-Limits.md:7]:
```suggestion
You can think of a `product` rate limit like a "bucket". Therefore, when a request is made to a `product` (which corresponds directly to an API endpoint), a single gallon of water is added to that endpoint’s bucket, thus starting that bucket's 60-second time-to-live (TTL) timer.
```

Comment (alexromano) [reply docs/Development-Guides/Rate-Limits.md:7]: nit: tokens or balls in a bucket might be more clear

File: docs/Development-Guides/Rate-Limits.md [hunk 7]
```diff
@@ -1,25 +1,18 @@
...
 
-Applications are rate limited on a per `product` basis with the following limits.
+After the product's rate limit is reset after 60 seconds, the first request to that `product` starts the 60-second Time-To-Live (TTL) timer again.
```
Comment (alexromano) [inline docs/Development-Guides/Rate-Limits.md:9]:
```suggestion
After the product's rate limit is reset after 60 seconds, the first request to that `product` starts the 60-second TTL again.
```
nit: after you define the acronym once you can just use it thereafter

File: docs/Development-Guides/Rate-Limits.md [hunk 8]
```diff
@@ -1,25 +1,18 @@
...
-`employment` | 20
-`payment` | 12
-`pay-statement` | 12
+***
 
-## Access Token
+## Access Token Rate Limits
 
-Access tokens are rate limited on a per `product` basis with the following limits.
+For access token rate limits, each `product` manages its own "bucket" with its own 60-second Time-To-Live (TTL) timer. The first request by an access token to a new product endpoint starts the timer for only that product's bucket, and rate limits reset only for that product endpoint after the 60-second period.
```
Comment (alexromano) [inline docs/Development-Guides/Rate-Limits.md:15]:
```suggestion
For access token rate limits, each product has its own "bucket" with its own 60-second TTL timer. The first request made by an access token to a new product endpoint starts the timer for that specific product's bucket. Rate limits reset only for that product endpoint after the 60-second period.
```

File: docs/Development-Guides/Rate-Limits.md [hunk 9]
```diff
@@ -29,3 +22,74 @@ Product | Max requests initiated per minute
...
+    “code”: 429,
+    “message”: “Too many requests for token”,
+    “name”: “rate_limit_exceeded_error”,
+    “finch_code”: “finch_token_rl”
+}
+```
+
+## Application Rate Limits
+
+Similarly, multiple access tokens can be created from a single Finch application as more employers are connected. A Finch application has its own rate limits separate from the access token rate limits. You can think of an application like a development environment or application stage. Finch will assign three types of applications: `sandbox`, `development`, and `production`. Each will have their own `client_id` which represents the application.
```
Comment (alexromano) [inline docs/Development-Guides/Rate-Limits.md:42]:
```suggestion
Similarly, multiple access tokens can be created from a single Finch application as more employers are connected. A Finch application has its own rate limits separate from the access token rate limits. (A Finch application corresponds to a unique `client_id`. You may have several `client_id`s if you use a development or sandbox application in addition to production).
```
I think we can find a middle ground and just be more concrete/succinct here.

File: docs/Development-Guides/Rate-Limits.md [hunk 10]
```diff
@@ -29,3 +22,74 @@ Product | Max requests initiated per minute
...
+    "status": 429,
+    "code”: 429,
+    "message": "Too many requests for token",
+    "name": "rate_limit_exceeded_error",
+    "finch_code": "finch_api_ip_rl"
+}
+```
+
+***
+
```
Comment (alexromano) [inline docs/Development-Guides/Rate-Limits.md:92]: do we want to call out anywhere here that in addition to these rate limits, we can also return 429s if an upstream provider rate limits us?

File: docs/Best-Practices/Handling-Rate-Limit-Errors.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
+# Handling Rate Limit Errors
+
+If you are experiencing rate limits, there are several ways to minimize the risk of hitting rate limits.
+
+## Implement batched requests
+
+Minimize the risk of hitting rate limits by batching individual requests into a single request. Several of Finch’s endpoints are “batch” endpoints ([/individual](https://developer.tryfinch.com/docs/reference/9d6c83b09e205-individual), [/employment](https://developer.tryfinch.com/docs/reference/1ba5cdec4c979-employment), and [/pay-statment](https://developer.tryfinch.com/docs/reference/d5fd02c41e83a-pay-statement)). Batch endpoints allow a list of IDs to be sent in a single request. Finch will correspondingly return a single response with an array of objects equal to the IDs sent.
```
Comment (alexromano) [inline docs/Best-Practices/Handling-Rate-Limit-Errors.md:7]:
```suggestion
Minimize the risk of hitting rate limits by batching individual requests into a single request. Several of Finch’s endpoints are “batch” endpoints ([/individual](https://developer.tryfinch.com/docs/reference/9d6c83b09e205-individual), [/employment](https://developer.tryfinch.com/docs/reference/1ba5cdec4c979-employment), and [/pay-statment](https://developer.tryfinch.com/docs/reference/d5fd02c41e83a-pay-statement)). These batch endpoints allow you to send a list of IDs in a single request and receive a response with an array of objects matching the IDs sent.
```

File: docs/Best-Practices/Handling-Rate-Limit-Errors.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  "requests": [
+    {
+      "individual_id": "772b3c4f-d764-433d-bd69-ff8bbac33ffe"
+    },
+    {
+      "individual_id": "a7a77065-0f68-418d-a85a-da24fb2139b7"
+    },
+    .
+    .
+    .
```
Comment (alexromano) [inline docs/Best-Practices/Handling-Rate-Limit-Errors.md:24]:
```suggestion
    ...
```

File: docs/Best-Practices/Handling-Rate-Limit-Errors.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    {
+      "individual_id": "84364585-c2ce-40aa-bcc0-666ac9577315"
+    }
+  ]
+}
+```
+
+## Implement a Rate Limiter Enforcer
+
+Suppose the rate limit for a particular Finch `product` is 20 requests per minute. You can use the following JavaScript code example to enforce this rate limit quota in your application. The `RateLimiter` class implements a simple rate limiter that allows you to make requests up to the specified rate limit (20 requests per minute) and pauses further requests until the rate limit resets before resuming API requests. Alternatively to pausing, you can implement a “back-off and retry” strategy. The `request` method of the rate limiter is used to make API requests to Finch's endpoints, ensuring that you stay within the rate limit "bucket" quota for that endpoint. Simply initialize a new `RateLimiter` class for each Finch `product` endpoint being called.
```
Comment (alexromano) [inline docs/Best-Practices/Handling-Rate-Limit-Errors.md:34]:
```suggestion
Suppose the rate limit for a particular Finch `product` is 20 requests per minute. You can use the following JavaScript code example to enforce this rate limit quota in your application. The `RateLimiter` class implements a simple rate limiter that allows you to make requests up to the specified rate limit (20 requests per minute) and pauses further requests until the rate limit resets . As an alternative, you can implement a “back off and retry” strategy. The `request` method of the rate limiter is used to make API requests to Finch's endpoints, ensuring that you stay within the rate limit "bucket" quota for that endpoint. Simply initialize a new `RateLimiter` class for each Finch `product` endpoint being called.
```

File: docs/Best-Practices/Handling-Rate-Limit-Errors.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+    - Application-level rate limits
+        Bucket | Capacity
+        -------|-------------
+        `company` | 4/20
+        `directory` | 3/20
+        `individual` | 0/20
+        `employment` | 0/20
+        `payment` | 2/12
+        `pay-statement` | 0/12
+1. Token B makes 5 requests to the [/employer/company](https://developer.tryfinch.com/docs/reference/33162be1eed72-company) endpoint, 3 requests to the [/employer/directory](https://developer.tryfinch.com/docs/reference/12419c085fc0e-directory) endpoint, and 2 requests to the [/employer/payment](https://developer.tryfinch.com/docs/reference/b811fdc2542ca-payment) endpoint within the same minute. Token B’s first 4 requests to the [/employer/company](https://developer.tryfinch.com/docs/reference/33162be1eed72-company) endpoint will succeed, but the 5th will fail and return a 429 rate limit error. Note: Only succeeded requests count towards the application level rate limits.
```
Comment (alexromano) [inline docs/Best-Practices/Handling-Rate-Limit-Errors.md:150]: each number here is "1". not sure if that's intentional

**Review by alexromano — APPROVED** [review]
this is great, thanks for doing this!

File: toc.json [hunk 15]
```diff
@@ -188,10 +193,10 @@
...
     },
-     {
+    {
       "type": "item",
       "title": "ATS",
       "uri": "reference/ats/openapi.yaml"
     }
   ]
-}
+}
```
[the remark this answers is not in the document]
Comment (tylernix) [reply toc.json]: done
